core/src/Streamly/Internal/FileSystem/Path.hs

```diff
@@ -1,515 +1,194 @@
--- On Posix it may fail if the stream contains null characters.
--- TBD: Use UTF16LE on Windows.
-fromString :: MonadThrow m => [Char] -> m Path
-fromString = fromChars . Stream.fromList
-
--- | Decode the path to a Unicode string using strict UTF-8 decoding on Posix.
--- TBD: Use UTF16LE on Windows.
-toString :: Path -> [Char]
-toString = runIdentity . Stream.toList . toChars
-
-------------------------------------------------------------------------------
--- Statically Verified Literals
-------------------------------------------------------------------------------
-
--- XXX Build these on top of the str quasiquoter so that we get the
--- interpolation for free.
-
--- | Generates a 'Path' type from an interpolated string literal.
+--  * https://en.wikipedia.org/wiki/Path_(computing)
+--  * https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file
+--  * https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/62e862f4-2a51-452e-8eeb-dc4ff5ee33cc
 --
--- /Unimplemented/
-path :: QuasiQuoter
-path = undefined
-
--- | Generates an @Abs Path@ type from an interpolated string literal.
+-- == File System Tree
 --
--- /Unimplemented/
-abs :: QuasiQuoter
-abs = undefined
-
--- | Generates an @Rel Path@ type from an interpolated string literal.
+-- A filesystem is a tree when there are no hard links or symbolic links. But
+-- in the presence of symlinks it could be a DAG or a graph, because directory
+-- symlinks can create cycles.
 --
--- /Unimplemented/
-rel :: QuasiQuoter
-rel = undefined
-
--- | Generates an @Dir Path@ type from an interpolated string literal.
+-- == Location and Segments
 --
--- /Unimplemented/
-dir :: QuasiQuoter
-dir = undefined
-
--- | Generates an @File Path@ type from an interpolated string literal.
+-- We make two distinctions for paths, a path could refer to a location or it
+-- could refer to a segment or segments.
 --
--- /Unimplemented/
-file :: QuasiQuoter
-file = undefined
-
--- | Generates an @Abs (Dir Path)@ type from an interpolated string literal.
+-- A path that refers to a particular object in the file system  is called a
+-- location e.g. /usr is a location, . is a location, ./bin is a location. A
+-- location could be absolute e.g. /usr or it could be relative e.g. ./bin . A
+-- location always has two components, a specific "root" which could be
+-- explicit or implicit, and a path segment relative to the root. A location
+-- with a fixed root is known as an absolute location whereas a location with
+-- an implicit root e.g. "./bin" is known as a relative location.
 --
--- /Unimplemented/
-absdir :: QuasiQuoter
-absdir = undefined
-
--- | Generates an @Rel (Dir Path)@ type from an interpolated string literal.
+-- A path that does not refer to a particular location but defines steps to go
+-- from some place to another is a path segment. For example, "local/bin" is a
+-- path segment whereas "./local/bin" is a location.
```

Comment:
Ah, in the above doc, `../bin` is given as an example for a segment.
Why can't `./local/bin` be a segment?
Is `.` treated differently from `..`?

Reply:
`./local/bin` points to a specific place i.e. `local/bin` in the current directory. `../bin` is not relative to anything, it is a segment but `./../bin` would be a location as it is relative to current directory.